### PR TITLE
feat(bootstrap): ✨ add struct field access + construction (Tier B)

### PR DESCRIPTION
**Pull request by justapithecus**
## Summary

Tier B slice: full struct type support across the bootstrap MIR and LLVM backend. Struct field reads (`p.x`), struct construction (`P(1, 2)`), struct-typed call arguments, struct-typed return values, and LLVM struct type definitions all work end-to-end with verified clang compilation.

### MIR (`bootstrap/mir/impl.dao`)
- **`mir_resolve_field_index`**: walks `DaoType.type_info` field pairs to resolve `field_index` from `-1` to the actual positional index during `HirFieldAccess` lowering

### LLVM backend (`bootstrap/llvm/impl.dao`)
- **Struct type definitions** emitted at module top: `%struct.P = type { i32, i32 }`
- **`LlStructDef`** tracking + `bs_ensure_struct_def` for deduped struct def registration
- **`LlExtractValue`** for struct field reads — works on SSA struct values directly (GEP would need a pointer, but Task 29 MIR gives us a loaded struct value from `MirLoad`)
- **`LlGepField`** kept for the struct constructor path (where we do have a pointer from alloca)
- **Struct constructor recognition**: `MirCall` on `SymbolKind.Type` → alloca + per-field GEP stores + load of the complete struct value
- **Struct-aware type text everywhere**: `LlAlloca`, `LlStore`, `LlLoad`, `LlParam`, `LlCall.ret_ty_text`, `LlRet.value_ty_text`, `LlFunction.ret_ty_text`, and `CallArgs.arg_ty_texts` all carry type text strings so `%struct.P` serializes correctly at call and return edges (previously produced `<unsupported>`)
- **`type_info`** added to `MirBackendInput` and `BS` for struct field type lookup
- **`bs_type_text`** helper returns `%struct.Name` for struct types

### Latent Task 30 SSA bug fixed

`bs_fresh_value` mutated `bs.next_value` on a value-copied local parameter; the mutation never propagated back to the caller. Every call returned `"%0"`, producing invalid SSA with duplicate names. Task 30's substring-only tests never caught this — they checked for things like `"mul i32 2, 3"` without verifying SSA value names. Fix: `bs_fresh_value` now returns a `FreshR { bs, name }` pair and all 10 call sites rebind `bs` from the return.

### New edge-case tests (now 16 LLVM tests, was 12)

- `struct_field_read`: `fn get_x(p: P): i32 → p.x` emits `%struct.P = type { i32, i32 }` + `extractvalue %struct.P`
- `struct_construction`: `let p = P(1, 2)` emits alloca + GEP stores per field
- `struct_call_arg`: passes a struct to a function — checks `define i32 @take(%struct.P` + `call i32 @take(%struct.P` + no `<unsupported>` marker
- `struct_return`: function returning a struct — checks `define %struct.P @make()` + `ret %struct.P` + `call %struct.P @make()` + no `<unsupported>` marker

### Manual clang validation

```
$ cat /tmp/dao_struct_return.ll
%struct.P = type { i32, i32 }
define %struct.P @make() {
entry:
  %0 = alloca %struct.P
  %1 = getelementptr inbounds %struct.P, ptr %0, i32 0, i32 0
  store i32 5, ptr %1
  %2 = getelementptr inbounds %struct.P, ptr %0, i32 0, i32 1
  store i32 6, ptr %2
  %3 = load %struct.P, ptr %0
  ret %struct.P %3
}
define i32 @main() {
entry:
  %slot0 = alloca %struct.P
  %0 = call %struct.P @make()
  store %struct.P %0, ptr %slot0
  %1 = load %struct.P, ptr %slot0
  %2 = extractvalue %struct.P %1, 0
  ret i32 %2
}

$ clang /tmp/dao_struct_return.ll -o /tmp/dao_struct_return && /tmp/dao_struct_return; echo \$?
5
```

Also validated `let x = 1 + 2 * 3; return x` → exit code 7 (catches the SSA bug fix).

## Test plan

- [x] Typecheck: 43/43
- [x] HIR: 22/22
- [x] MIR: 8/8
- [x] **LLVM: 16/16** (was 13, +2 struct tests +1 struct_construction kept)
- [x] **291/291 bootstrap tests passing**
- [x] Manual clang validation of struct-accessing IR: exit code 5
- [x] Manual clang validation of SSA fix (arithmetic): exit code 7

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bootstrap/llvm/impl.dao [hunk 1]
```diff
@@ -796,11 +909,35 @@ fn bs_lower_inst(bs: BS, insts: Vector<LlInst>, mir_node_idx: i64): LowerStepR
...
+      let obj_ty_text: string = bs_get_value_ty(bs, obj)
+      let obj_text: string = bs_get_value_text(bs, obj)
+      // obj is a pointer to the struct (alloca'd).  GEP into the
+      // field at field_index, then load the field value.
+      let gep_result: string = bs_fresh_value(bs)
+      let field_ty: LlType = ll_type_from_mir(bs.types, type_idx)
+      let field_ty_text: string = bs_type_text(bs.types, type_idx)
+      // Emit: %gep = getelementptr <struct_ty>, ptr <obj>, i32 0, i32 <field_index>
+      let gep_inst: LlInst = LlInst(
+        LlInstKind.LlGepField(obj_text, obj_ty_text, field_index),
```
[the remark this answers is not in the document]
Comment (justapithecus) [reply bootstrap/llvm/impl.dao]: Fixed in d6066be before this review thread landed. The GEP+load path for `MirFieldAccess` was replaced with `extractvalue`, which works on SSA struct values directly (no pointer required). The inline code now reads:

```
LlInstKind.LlExtractValue(obj_text, obj_ty_text, field_index)
```

GEP is retained only for the struct constructor path, where we do have a pointer from the preceding alloca. Verified end-to-end with clang: struct_return → exit code 5.

File: bootstrap/llvm/impl.dao [hunk 2]
```diff
@@ -456,6 +498,30 @@ fn bs_mark_extern(bs: BS, name: string): BS
...
+  // Build the struct def from the DaoType's field info.
+  let t: DaoType = bs.types.get(type_idx)
+  let ll_name: string = "%struct." + t.name
+  let field_tys: Vector<string> = Vector<string>::new()
+  let fi: i64 = to_i64(0)
+  while fi < t.info_count:
+    // type_info stores (name_tidx, field_type_idx) pairs starting
+    // at t.info_lp with t.info_count pairs.
+    let field_type_idx: i64 = bs.type_info.get(t.info_lp + fi * to_i64(2) + to_i64(1))
+    field_tys = field_tys.push(bs_type_text(bs.types, field_type_idx))
```
[the remark this answers is not in the document]
Comment (justapithecus) [reply bootstrap/llvm/impl.dao:570]: Fixed in 20a3286. `bs_ensure_struct_def` now walks the struct's field types and recursively registers any struct field type it encounters. Cycle-safe via dedup-first-then-recurse — a placeholder entry is pushed before recursing so self/mutual references terminate, and the placeholder's `field_tys` is updated afterwards via `bs.struct_defs.set(slot_idx, ...)`.

New test `nested_struct_registration`: `class Outer:\n  inner: Inner\n` now emits both `%struct.Inner = type { i32 }` and `%struct.Outer = type { %struct.Inner }`. 17/17 LLVM backend tests pass.

File: bootstrap/llvm/impl.dao [hunk 3]
```diff
@@ -179,6 +187,9 @@ fn ll_type_text(ty: LlType): string
       return "void"
     LlType.LlPtr:
       return "ptr"
+    LlType.LlStruct:
+      // Struct name requires the type_idx → use bs_type_text instead.
+      return "<struct>"
```
[the remark this answers is not in the document]
Comment (justapithecus) [reply bootstrap/llvm/impl.dao]: Fixed in 20a3286. The earlier type-model refactor (commit 07cdbb5) removed all the call sites where a struct `LlType` could reach `ll_type_text` — `LlInst.result_ty`, `LlParam.ty`, and `LlFunction.ret_ty` are gone, and serialization reads from the parallel `*_ty_text` fields exclusively. So the leak path is no longer reachable in practice.

To prevent regressions, `ll_type_text(LlStruct)` now returns `"<bug:ll_type_text-of-LlStruct-use-bs_type_text>"` instead of `"<struct>"`. Any accidental future caller shows up immediately in emitted IR and in substring tests. The doc comment at the top of `ll_type_text` explains that callers must use `bs_type_text(types, type_idx)` for any type_idx that might be a struct, because `LlStruct` carries no name of its own.
